<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Tools for tidy metagenomics • metaglomr</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Tools for tidy metagenomics">
<meta property="og:description" content="Aggregates count data, sample metadata, and taxonomic annotations into long-form dataframes.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">metaglomr</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/eclarke/metaglomr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    

    
    
<p>This is a package to combine counts, sample metadata, and taxonomic annotations into one long-form tidy dataframe for further analysis. It offers functions to switch between phyloseq objects and these ‘agglomerated’ dataframes.</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="st">"eclarke/metaglomr"</span>)</code></pre></div>
</div>
<div id="short-usage-overview" class="section level2">
<h2 class="hasAnchor">
<a href="#short-usage-overview" class="anchor"></a>Short usage overview</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(tidyverse))
<span class="kw">library</span>(metaglomr)

<span class="kw">data</span>(features)  <span class="co"># A counts matrix, with rows as samples and columns as features</span>
<span class="kw">data</span>(samples)   <span class="co"># A dataframe with sample metadata</span>
<span class="kw">data</span>(taxa)      <span class="co"># A dataframe or matrix with taxonomic annotations for features</span>

<span class="co"># Combine all of these datasets into one long-form dataframe</span>
(agg &lt;-<span class="st"> </span><span class="kw"><a href="reference/agglomerated.html">agglomerated</a></span>(features, samples, taxa, <span class="st">"sample_id"</span>))</code></pre></div>
<pre><code>## # A tibble: 1,024 x 12
##    sample_id otu_id count study_group sample_type Kingdom  Phylum   Class 
##    &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt; &lt;fct&gt;       &lt;fct&gt;       &lt;fct&gt;    &lt;fct&gt;    &lt;fct&gt; 
##  1 S_1       f_1       25 case        A           Bacteria Bactero… Bacte…
##  2 S_2       f_1       78 case        B           Bacteria Bactero… Bacte…
##  3 S_3       f_1       34 case        C           Bacteria Bactero… Bacte…
##  4 S_4       f_1       61 case        D           Bacteria Bactero… Bacte…
##  5 S_5       f_1      258 case        A           Bacteria Bactero… Bacte…
##  6 S_6       f_1       54 case        B           Bacteria Bactero… Bacte…
##  7 S_7       f_1       56 case        C           Bacteria Bactero… Bacte…
##  8 S_8       f_1       58 case        D           Bacteria Bactero… Bacte…
##  9 S_9       f_1       48 control     A           Bacteria Bactero… Bacte…
## 10 S_10      f_1        7 control     B           Bacteria Bactero… Bacte…
## # ... with 1,014 more rows, and 4 more variables: Order &lt;fct&gt;,
## #   Family &lt;fct&gt;, Genus &lt;fct&gt;, Species &lt;fct&gt;</code></pre>
</div>
<div id="motivation" class="section level2">
<h2 class="hasAnchor">
<a href="#motivation" class="anchor"></a>Motivation</h2>
<p>In metagenomics, we often work with three distinct datasets: a feature count table, a sample table, and a taxonomic annotation table.</p>
<div id="feature-count-table" class="section level3">
<h3 class="hasAnchor">
<a href="#feature-count-table" class="anchor"></a>Feature count table</h3>
<p>The feature count table is is a matrix of features <span class="math inline">\(\times\)</span> samples, and the cells of the matrix are the times that feature was seen in that sample. Features can be OTUs, ASVs, species, genomes, or whatever. Often these come from a biom file.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(features)
features[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">15</span>]</code></pre></div>
<pre><code>##     f_1 f_2 f_3 f_4 f_5 f_6 f_7 f_8 f_9 f_10 f_11 f_12 f_13 f_14 f_15
## S_1  25 112 130 123 101  61 220  72  43   19  105   97   52   76  258
## S_2  78 402  67 188  69  89 150  44 129  236   56   83  284   15  112
## S_3  34 728 293 342   4   6  19 122 110   19  387    5  116   17  148
## S_4  61  84  41   5 106  39  97  51  97   58  322   84  132    1   35
## S_5 258  40  65 160  38  31  52  88  36   88   41   91  128  140  105</code></pre>
</div>
<div id="sample-metadata" class="section level3">
<h3 class="hasAnchor">
<a href="#sample-metadata" class="anchor"></a>Sample metadata</h3>
<p>The sample metadata can be anything about the samples, but frequently contains at least a description of the sample type and study group of each sample. This is frequently referred to as a mapping file in QIIME.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(samples)
<span class="kw">as_tibble</span>(samples)</code></pre></div>
<pre><code>## # A tibble: 16 x 3
##    sample_id study_group sample_type
##    &lt;fct&gt;     &lt;fct&gt;       &lt;fct&gt;      
##  1 S_1       case        A          
##  2 S_2       case        B          
##  3 S_3       case        C          
##  4 S_4       case        D          
##  5 S_5       case        A          
##  6 S_6       case        B          
##  7 S_7       case        C          
##  8 S_8       case        D          
##  9 S_9       control     A          
## 10 S_10      control     B          
## 11 S_11      control     C          
## 12 S_12      control     D          
## 13 S_13      control     A          
## 14 S_14      control     B          
## 15 S_15      control     C          
## 16 S_16      control     D</code></pre>
</div>
<div id="taxonomic-annotations" class="section level3">
<h3 class="hasAnchor">
<a href="#taxonomic-annotations" class="anchor"></a>Taxonomic annotations</h3>
<p>This contains taxonomic annotations for all the features (OTUs or otherwise), split by rank (i.e. Kingdom, Phylum, etc).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(taxa)
taxa[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, ]</code></pre></div>
<pre><code>##      Kingdom         Phylum               Class            Order
## f_1 Bacteria  Bacteroidetes         Bacteroidia    Bacteroidales
## f_2 Bacteria Proteobacteria Alphaproteobacteria      Rhizobiales
## f_3 Bacteria     Firmicutes       Negativicutes  Selenomonadales
## f_4 Bacteria     Firmicutes          Clostridia    Clostridiales
## f_5 Bacteria  Bacteroidetes         Bacteroidia Flavobacteriales
##                 Family                     Genus Species
## f_1 Porphyromonadaceae             Porphyromonas    &lt;NA&gt;
## f_2  Xanthobacteraceae              Pseudolabrys    &lt;NA&gt;
## f_3    Veillonellaceae               Veillonella    &lt;NA&gt;
## f_4  Defluviitaleaceae Defluviitaleaceae_UCG-011    &lt;NA&gt;
## f_5      Weeksellaceae          Chryseobacterium    &lt;NA&gt;</code></pre>
</div>
<div id="agglomerating-the-three-datasets" class="section level3">
<h3 class="hasAnchor">
<a href="#agglomerating-the-three-datasets" class="anchor"></a>Agglomerating the three datasets</h3>
<p>For users of the <a href="https://tidyverse.org">tidyverse</a>, it’s frequently easiest to work with long-form melted datasets, where each row is a unique observation or data point. The unique reference in these three datasets is the count of a feature in a particular sample. Therefore, we can create a dataframe where each row is this unique combination of feature + sample + count, with additional columns describing the sample and feature further.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">(agg &lt;-<span class="st"> </span><span class="kw"><a href="reference/agglomerated.html">agglomerated</a></span>(features, samples, taxa, <span class="st">"sample_id"</span>))</code></pre></div>
<pre><code>## # A tibble: 1,024 x 12
##    sample_id otu_id count study_group sample_type Kingdom  Phylum   Class 
##    &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt; &lt;fct&gt;       &lt;fct&gt;       &lt;fct&gt;    &lt;fct&gt;    &lt;fct&gt; 
##  1 S_1       f_1       25 case        A           Bacteria Bactero… Bacte…
##  2 S_2       f_1       78 case        B           Bacteria Bactero… Bacte…
##  3 S_3       f_1       34 case        C           Bacteria Bactero… Bacte…
##  4 S_4       f_1       61 case        D           Bacteria Bactero… Bacte…
##  5 S_5       f_1      258 case        A           Bacteria Bactero… Bacte…
##  6 S_6       f_1       54 case        B           Bacteria Bactero… Bacte…
##  7 S_7       f_1       56 case        C           Bacteria Bactero… Bacte…
##  8 S_8       f_1       58 case        D           Bacteria Bactero… Bacte…
##  9 S_9       f_1       48 control     A           Bacteria Bactero… Bacte…
## 10 S_10      f_1        7 control     B           Bacteria Bactero… Bacte…
## # ... with 1,014 more rows, and 4 more variables: Order &lt;fct&gt;,
## #   Family &lt;fct&gt;, Genus &lt;fct&gt;, Species &lt;fct&gt;</code></pre>
<p>While this may seem overly repetitive (as the metadata is duplicated in lots of rows), R and dplyr actually handle this pretty well. Things only start breaking down with feature tables that have more than &gt; 100,000,000 cells. What this buys you is the ability to use standard tidyverse verbs and operations easily.</p>
</div>
</div>
<div id="examples" class="section level2">
<h2 class="hasAnchor">
<a href="#examples" class="anchor"></a>Examples</h2>
<div id="subsetting" class="section level3">
<h3 class="hasAnchor">
<a href="#subsetting" class="anchor"></a>Subsetting</h3>
<p>Subsetting is easy through the use of the <code>filter</code> verb:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">filter</span>(agg, study_group <span class="op">==</span><span class="st"> "case"</span>)</code></pre></div>
<pre><code>## # A tibble: 512 x 12
##    sample_id otu_id count study_group sample_type Kingdom  Phylum  Class  
##    &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt; &lt;fct&gt;       &lt;fct&gt;       &lt;fct&gt;    &lt;fct&gt;   &lt;fct&gt;  
##  1 S_1       f_1       25 case        A           Bacteria Bacter… Bacter…
##  2 S_2       f_1       78 case        B           Bacteria Bacter… Bacter…
##  3 S_3       f_1       34 case        C           Bacteria Bacter… Bacter…
##  4 S_4       f_1       61 case        D           Bacteria Bacter… Bacter…
##  5 S_5       f_1      258 case        A           Bacteria Bacter… Bacter…
##  6 S_6       f_1       54 case        B           Bacteria Bacter… Bacter…
##  7 S_7       f_1       56 case        C           Bacteria Bacter… Bacter…
##  8 S_8       f_1       58 case        D           Bacteria Bacter… Bacter…
##  9 S_1       f_2      112 case        A           Bacteria Proteo… Alphap…
## 10 S_2       f_2      402 case        B           Bacteria Proteo… Alphap…
## # ... with 502 more rows, and 4 more variables: Order &lt;fct&gt;, Family &lt;fct&gt;,
## #   Genus &lt;fct&gt;, Species &lt;fct&gt;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">filter</span>(agg, Phylum <span class="op">==</span><span class="st"> "Bacteroidetes"</span>)</code></pre></div>
<pre><code>## # A tibble: 176 x 12
##    sample_id otu_id count study_group sample_type Kingdom  Phylum   Class 
##    &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt; &lt;fct&gt;       &lt;fct&gt;       &lt;fct&gt;    &lt;fct&gt;    &lt;fct&gt; 
##  1 S_1       f_1       25 case        A           Bacteria Bactero… Bacte…
##  2 S_2       f_1       78 case        B           Bacteria Bactero… Bacte…
##  3 S_3       f_1       34 case        C           Bacteria Bactero… Bacte…
##  4 S_4       f_1       61 case        D           Bacteria Bactero… Bacte…
##  5 S_5       f_1      258 case        A           Bacteria Bactero… Bacte…
##  6 S_6       f_1       54 case        B           Bacteria Bactero… Bacte…
##  7 S_7       f_1       56 case        C           Bacteria Bactero… Bacte…
##  8 S_8       f_1       58 case        D           Bacteria Bactero… Bacte…
##  9 S_9       f_1       48 control     A           Bacteria Bactero… Bacte…
## 10 S_10      f_1        7 control     B           Bacteria Bactero… Bacte…
## # ... with 166 more rows, and 4 more variables: Order &lt;fct&gt;, Family &lt;fct&gt;,
## #   Genus &lt;fct&gt;, Species &lt;fct&gt;</code></pre>
</div>
<div id="adding-new-columns" class="section level3">
<h3 class="hasAnchor">
<a href="#adding-new-columns" class="anchor"></a>Adding new columns</h3>
<p>Here’s how to calculate proportions and add it to your dataframe:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">agg &lt;-<span class="st"> </span>agg <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">group_by</span>(sample_id) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">proportion =</span> count<span class="op">/</span><span class="kw">sum</span>(count))

<span class="co"># Showing just a subset of the data:</span>
<span class="kw">select</span>(agg, sample_id, otu_id, count, proportion)</code></pre></div>
<pre><code>## # A tibble: 1,024 x 4
## # Groups:   sample_id [16]
##    sample_id otu_id count proportion
##    &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt;      &lt;dbl&gt;
##  1 S_1       f_1       25    0.00429
##  2 S_2       f_1       78    0.0104 
##  3 S_3       f_1       34    0.00510
##  4 S_4       f_1       61    0.0101 
##  5 S_5       f_1      258    0.0434 
##  6 S_6       f_1       54    0.00811
##  7 S_7       f_1       56    0.00874
##  8 S_8       f_1       58    0.0120 
##  9 S_9       f_1       48    0.00635
## 10 S_10      f_1        7    0.00116
## # ... with 1,014 more rows</code></pre>
</div>
<div id="aggregating-based-on-metadata" class="section level3">
<h3 class="hasAnchor">
<a href="#aggregating-based-on-metadata" class="anchor"></a>Aggregating based on metadata:</h3>
<p>Aggregate based on taxonomic rank:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">agg <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">group_by</span>(sample_id, Phylum) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">count =</span> <span class="kw">sum</span>(count)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ungroup</span>() <span class="op">%&gt;%</span>
<span class="st">  </span><span class="co"># re-add sample data that got lost in the summarizing</span>
<span class="st">  </span><span class="kw">left_join</span>(<span class="kw"><a href="reference/get_samples.html">get_samples</a></span>(agg, sample_id, study_group, sample_type))</code></pre></div>
<pre><code>## Joining, by = "sample_id"</code></pre>
<pre><code>## # A tibble: 80 x 5
##    sample_id Phylum         count study_group sample_type
##    &lt;chr&gt;     &lt;fct&gt;          &lt;dbl&gt; &lt;fct&gt;       &lt;fct&gt;      
##  1 S_1       Actinobacteria   363 case        A          
##  2 S_1       Bacteroidetes   1007 case        A          
##  3 S_1       Firmicutes      3616 case        A          
##  4 S_1       Proteobacteria   699 case        A          
##  5 S_1       Spirochaetes     136 case        A          
##  6 S_10      Actinobacteria   627 control     B          
##  7 S_10      Bacteroidetes   1169 control     B          
##  8 S_10      Firmicutes      2869 control     B          
##  9 S_10      Proteobacteria  1043 control     B          
## 10 S_10      Spirochaetes     323 control     B          
## # ... with 70 more rows</code></pre>
</div>
<div id="summarizing" class="section level3">
<h3 class="hasAnchor">
<a href="#summarizing" class="anchor"></a>Summarizing</h3>
<p>Or find the most prevalent phyla in your study groups:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">agg <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">ungroup</span>() <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">group_by</span>(study_group, Phylum) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">mean_proportion =</span> <span class="kw">mean</span>(proportion)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">top_n</span>(<span class="dv">1</span>, mean_proportion)</code></pre></div>
<pre><code>## # A tibble: 2 x 3
## # Groups:   study_group [2]
##   study_group Phylum        mean_proportion
##   &lt;fct&gt;       &lt;fct&gt;                   &lt;dbl&gt;
## 1 case        Firmicutes             0.0172
## 2 control     Bacteroidetes          0.0173</code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/eclarke/metaglomr">https://​github.com/​eclarke/​metaglomr</a>
</li>
<li>Report a bug at <br><a href="https://github.com/eclarke/metaglomr/issues">https://​github.com/​eclarke/​metaglomr/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Erik Clarke <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>

      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Erik Clarke.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
